desk/lib/discipline.hoon

```diff
@@ -0,0 +1,307 @@
+      ~&  >>>  [%discipline dap=dap.bowl %bad-fact-mark n]
+      `(tell:log %crit ~['discipline: bad fact mark' >n<] ~)
+    ?~  type=(~(get by marks) p.cage.p.card)
+      ::NOTE  if we don't have a type for this mark,
+      ::      ignore it, since we can't test it
+      ~
+    ?:  (~(nest ut u.type) | p.q.cage.p.card)
+      ~
+    =+  n=[on-path=path with-mark=p.cage.p.card %nest-fail]
+    ~&  >>>  [%discipline dap=dap.bowl %bad-fact-data n]
+    `(tell:log %crit ~['discipline: bad fact data' >n<] ~)
+  ::  +check-scry: validate result against known marks
+  ::
+  ++  check-scry
+    |=  [=path result=(unit (unit cage))]
+    ^-  ?
+    ?.  ?=([~ ~ *] result)  &
+    ?~  want=(expected-scry-mark path)
+      ~&  [%discipline dap=dap.bowl %unknown-scry-path path=path mark=p.u.u.result]
+      &
+    ?:  =(%$ u.want)
+      ::NOTE  easy-out
+      &
+    ?.  =(p.u.u.result u.want)
+      ~&  >>>  [%discipline dap=dap.bowl %bad-scry-mark path=path want=want got=p.u.u.result]
+      |
+    ?~  type=(~(get by marks) p.u.u.result)
+      ::NOTE  if we don't have a type for this mark,
+      ::      ignore it, since we can't test it
+      &
+    ?:  (~(nest ut u.type) | p.q.u.u.result)
+      &
+    ~&  >>>  [%discipline dap=dap.bowl %bad-scry-data on-path=path with-mark=p.u.u.result %nest-fail]
+    |
+  ::  +check-marks: for marks where we know both sides, check equivalence
+  ::
+  ::    produces the marks that aren't equivalent
+  ::
+  ++  check-marks
+    |=  [ole=(map mark type) neu=(map mark [strict=? =type])]
+    ^-  (list mark)
+    %+  murn  ~(tap by ole)
+    |=  old=[=mark =type]
+    ^-  (unit mark)
+    ?~  new=(~(get by neu) mark.old)  ~
+    ?.  strict.u.new  ~
+    =;  match=?
+      ?:(match ~ `mark.old)
+    ?&  (~(nest ut type.old) | type.u.new)
+        (~(nest ut type.u.new) | type.old)
+    ==
+  --
+::
++$  state-0
+  $:  %0
+      last-marks=(map mark type)
+  ==
+::
+++  agent
+  =|  state-0
+  =*  state  -
+  ^-  agent:gall
+  !.  ::  we hide all the "straight into the inner agent" paths from traces
+  |_  =bowl:gall
+  +*  this    .
+      help    ~(. helper bowl last-marks)
+      og      ~(. inner inner-bowl:help)
+  ::
+  ++  on-init
+    ^-  (quip card _this)
+    =.  last-marks  mark-types
+    =^  cards  inner  on-init:og  !:
+    =.  cards  (check-cards:help ~ cards)
+    [cards this]
+  ::
+  ++  on-save  (slop !>([%discipline state]) on-save:og)
+  ++  on-load
+    |=  ole=vase
+    ^-  (quip card _this)
+    ?.  ?=([[%discipline *] *] q.ole)
+      =.  last-marks  mark-types
+      =^  cards  inner  (on-load:og ole)  !:
+      =.  cards  (check-cards:help ~ cards)
+      [cards this]
+    =+  !<([%discipline old=state-0] (slot 2 ole))
+    =.  state  old
+    ?^  bad=(check-marks:help last-marks mark-map)
+      ~|  [%discipline dap=dap.bowl %mark-types-changed marks=;;((list mark) bad)]
+      !!  ::REVIEW  possibly annoying during development?
```

Comment:
It would be annoying in heavy works (such as refactors), and would prompt us to just disable lib-discipline until types stabilize. If this proves to be annoying in other cases, perhaps a "development" flag would be welcome, but for now I would just leave it as is.

Reply:
The strictness flags on marks sort of already take care of this, so maybe can just remove the comment.